content/posts/cross-compile-query-json.mdx

```diff
@@ -0,0 +1,252 @@
+
+This post explains the project, how it was made and which decisions where followed and some reflections.
+
+## Why learning parsers/compilers
+I had a little idea about the theory, but very vague and useless. Since I create [styled-ppx](https://github.com/davesnx/styled-ppx) which is a
+ppx (**P**re**P**rocessor E**x**tension) that allows CSS-in-Reason/OCaml. Needs to parse the CSS and have a backend that compiles to [bs-emotion](https://github.com/ahrefs/bs-emotion).
+
+I asked for some help to [@EduardoRFS](https://www.twitch.tv/eduardorfs) about writting a CSS Parser that supports the entire [CSS3 specification](https://www.w3.org/TR/2001/WD-css3-roadmap-20010523/) which I will want to understand, improve and maintain over time.
+
+## How it works
+```bash
+query-json ".store.books | filter(.price > 10)" stores.json
+```
+
+This would read `stores.json` and access to `"store"` field, access to `"books"` field, since it's an array, it will run a filter on each item and if `"price" `field is bigger than 10 will keep that item on the list and finally print the resultant list.
+
+```json
+[
+  {
+    "title": "War and Peace",
+    "author": "Leo Tolstoy",
+    "price": 12.0
+  },
+  {
+    "title": "Lolita",
+    "author": "Vladimir Nabokov",
+    "price": 13.0
+  }
+]
+```
+
+The first argument it's called `query` it's a jq expression and the second one `json` or `payload` and it's a valid JSON file.
+
+The semantics of jq consists of a set of piped operations, where each output is connected to an input where the first input it's the JSON itself. Some pseudo-code to ilustate:
+
+```
+{ /* json */ } | filter | transform | count
+```
+
+In order to transform the `query` to a set of operations that run against a JSON, we will divide the problem in 3 steps: **parse**, **compile** and **run**.
+
+### Parsing
+Responsible for transforming a string into an Abstract Syntax Tree and provide a good error message when the input is malformed.
+
+<!-- TODO: Explain what kind of parser LR -->
+
+<!-- TODO: Explain this better/copy paste jq-language explaination -->
+One of the beauties of `jq` is that all the expressions are piped by default, so `.store | .books` is equivalent to `.store.books`. I designed the AST to represent the pipe structure in it's nature.
+
+Let's dive into an example. When the parser recieves `.store.books` will return:
+
+```reason
+Pipe(
+    Key("store"),
+    Key("books")
+)
+```
+
+All the operations are transformed to these constructors that you can see above (`Pipe`, `Key`).
+
+These expressions are called [Variants](https://reasonml.github.io/docs/en/variant), are one of the best features of Reason, some functional programming communities call it *ADT* (Algebraic Data Type) which is also correct.
+
+In this case those are *GADT* (Generic Algebraic Data Type) a kind of *ADTs* since they contain arguments and those can be recursive. The entire `query-json` program representation it's one big recursive Variant.
```

Comment:
```suggestion
In this case those are *GADT* (Generic Algebraic Data Type) a kind of *ADTs* since they contain arguments and those can be recursive. The entire `query-json` program representation is one big recursive Variant.
```